.github/workflows/Deploy.yaml

```diff
@@ -27,49 +31,101 @@ on:
+      directCommit: ${{ steps.CreateInputs.outputs.directCommit }}
+      requireEndToEndTests: ${{ steps.CreateInputs.outputs.requireEndToEndTests }}
+      createRelease: ${{ steps.CreateInputs.outputs.createRelease }}
+      defaultBcContainerHelperVersion: ${{ steps.CreateInputs.outputs.defaultBcContainerHelperVersion }}
+    steps:
+      - name: Create inputs
+        id: CreateInputs
+        run: |
+          if ($env:GITHUB_EVENT_NAME -eq 'workflow_run') {
+            $branch = 'preview'
+            $copyToMain = 'false'
+            $directCommit = 'false'
+            $requireEndToEndTests = 'true'
+            $createRelease = 'false'
+            $defaultBcContainerHelperVersion = 'preview'
+          } else {
+            $branch = '${{ github.event.inputs.branch }}'
+            $copyToMain = '${{ github.event.inputs.copyToMain }}'
+            $directCommit = '${{ github.event.inputs.directCommit }}'
+            $requireEndToEndTests = '${{ github.event.inputs.requireEndToEndTests }}'
+            $createRelease = '${{ github.event.inputs.createRelease }}'
+            $defaultBcContainerHelperVersion = '${{ github.event.inputs.defaultBcContainerHelperVersion }}'
+          }
+          Add-Content -encoding UTF8 -Path $env:GITHUB_OUTPUT -Value "branch=$branch"
+          Add-Content -encoding UTF8 -Path $env:GITHUB_OUTPUT -Value "copyToMain=$copyToMain"
+          Add-Content -encoding UTF8 -Path $env:GITHUB_OUTPUT -Value "directCommit=$directCommit"
+          Add-Content -encoding UTF8 -Path $env:GITHUB_OUTPUT -Value "requireEndToEndTests=$requireEndToEndTests"
+          Add-Content -encoding UTF8 -Path $env:GITHUB_OUTPUT -Value "createRelease=$createRelease"
+          Add-Content -encoding UTF8 -Path $env:GITHUB_OUTPUT -Value "defaultBcContainerHelperVersion=$defaultBcContainerHelperVersion"
+
   CheckEndToEnd:
     runs-on: [ ubuntu-latest ]
+    needs: [ Inputs ]
     steps:
       - name: Check successful end 2 end tests have run
-        if: github.repository_owner == 'microsoft' && github.event.inputs.requireEndToEndTests == 'true'
+        if: github.repository_owner == 'microsoft' && needs.Inputs.outputs.requireEndToEndTests == 'true'
         env:
           GH_TOKEN: ${{ github.token }}
         run: |
           $errorActionPreference = "Stop"
-          $end2endRuns = (gh api /repos/$($env:GITHUB_REPOSITORY)/actions/runs?per_page=100 | ConvertFrom-Json).workflow_runs | Where-Object { $_.name -eq 'End 2 End Tests' }
+          $end2endRuns = (gh api /repos/$($env:GITHUB_REPOSITORY)/actions/workflows/E2E.yaml/runs?per_page=100 | ConvertFrom-Json).workflow_runs
           $latestSha = (gh api /repos/$($env:GITHUB_REPOSITORY)/commits/main | ConvertFrom-Json).sha
-          $latestRun = $end2endruns | Where-Object { $_.head_sha -eq $latestSha }
+          $latestRun = $end2endruns | Where-Object { $_.Name -eq "End to end tests - $latestSha" } | select-object -first 1
```

Comment:
Shouldn't we check _all_ E2E tests run that were on the latest SHA?

What if I trigger E2E tests run because I didn't notice you ran it 10 minutes ago and mine fails due to an instability, but yours succeeds? Your workflow run should be good enough for deployment.

Reply:
I don't think so - I don't think we very very rarely will be in that situation.

I actually tried changing the code to check all - but then the code becomes more complex when trying to display which end 2 end test it is basing the verdict on. Now I can say; This end2end failed or this end2end is not complete - testing all means that I would say two has failed and one isn't complete???

Reply:
What about the following:
- If there is no E2E tests run on the latest SHA => throw "No E2E tests run found on SHA: _abcd_. Run E2E tests workflow."
- If none of the E2E tests runs on the latest SHA are "good" (rarely it has to check more than one run, but still possible) => "throw "None of the E2E tests runs on SHA _abcd_ were completed or successful. See runs at $runs.html_url"
- If there is a "good" E2E tests run => "Found successful E2E tests run on SHA abcd (link: $run.html_url)."

I imagine it gets complex because the script is inline. The alternative is a local action (I am a fan), so you can consider adding one for checking for successful E2E tests run. The other major pro of having an action instead of an inline script is that the code gets analyzed and all.

Reply:
Typically, there will be a few runs, where the last is the 2nd or 3rd attempt, which is successful, still running or failed.
Some times, there will be a partial end 2 end test which is successful (running some of the tests) - this should be excluded.
We have never (to this date and to my knowledge) been running two end 2 end tests at the same time.
We have never (to this date and to my knowledge) been running a partial end 2 end test on the same sha as an already successful full end 2 end test.
Also, if I run end 2 end tests personally - it is usually on the SHA of my PR (not on main).
So, I think you are trying to fix a problem which doesn't exist.

I do agree that we should refactor the internal workflows (including this) to not have inline code, but I won't have the time to do a major refactoring like this at this time - we can add a slice for that work.